state/export_test.go

```diff
@@ -183,29 +184,37 @@ func AddTestingCharmMultiSeries(c *gc.C, st *State, name string) *Charm {
 }
 
 func AddTestingApplication(c *gc.C, st *State, name string, ch *Charm) *Application {
-	return addTestingApplication(c, st, "", name, ch, nil, nil)
+	return addTestingApplication(c, st, "", name, ch, nil, nil, nil)
 }
 
 func AddTestingApplicationForSeries(c *gc.C, st *State, series, name string, ch *Charm) *Application {
-	return addTestingApplication(c, st, series, name, ch, nil, nil)
+	return addTestingApplication(c, st, series, name, ch, nil, nil, nil)
 }
 
 func AddTestingApplicationWithStorage(c *gc.C, st *State, name string, ch *Charm, storage map[string]StorageConstraints) *Application {
-	return addTestingApplication(c, st, "", name, ch, nil, storage)
+	return addTestingApplication(c, st, "", name, ch, nil, storage, nil)
+}
+
+func AddTestingApplicationWithDevices(c *gc.C, st *State, name string, ch *Charm, devices map[string]devices.Constraints) *Application {
+	return addTestingApplication(c, st, "", name, ch, nil, nil, devices)
 }
 
 func AddTestingApplicationWithBindings(c *gc.C, st *State, name string, ch *Charm, bindings map[string]string) *Application {
-	return addTestingApplication(c, st, "", name, ch, bindings, nil)
+	return addTestingApplication(c, st, "", name, ch, bindings, nil, nil)
 }
 
-func addTestingApplication(c *gc.C, st *State, series, name string, ch *Charm, bindings map[string]string, storage map[string]StorageConstraints) *Application {
+func addTestingApplication(
+	c *gc.C, st *State, series, name string, ch *Charm, bindings map[string]string,
```

Comment:
If a function has grown to  parameters, and a bunch of them are optional, we should definitely think about moving to a named parameter struct route. So it would be

type testingApplicationParams struct {
  c *gc.C
  st *State
...
}

func addTestingApplication(params testingApplicationParams)

and called as

addTestingApplication(testingApplicationParams{
  name: "foo",
  ch: mycharm,
})
etc.

I would be fine with something as ubiquitous as "c *gc.C" being passed in as its own argument, but once you're up at 6 params+ it is very hard to know which field is being passed in which location, and easy to get wrong. It also leads to helpers like the above, where you pass nil, or "" in a bunch of the places.

Reply:
yes, much nicer! thanks